### ISSUE TITLE
Missão 06

### ISSUE THREAD

**Issue by Adrielle-Araujo** [Adrielle-Araujo/MLCB#8]

**Comment by Adrielle-Araujo** [comment]
Adrielle Oliveira de Araujo
Victor Pereira de Souza

# PROJETO 1 - Classificação de Risco de Crédito

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

print("--- PROJETO 1: Classificação de Risco de Crédito ---")

# 1. Criando os dados
np.random.seed(42)
dados = pd.DataFrame({
    "idade": np.random.randint(25, 65, 100),
    "salario_anual": np.random.randint(30000, 200000, 100).astype(float),
    "anos_empregado": np.random.randint(0, 30, 100),
    "valor_emprestimo": np.random.randint(5000, 100000, 100),
    "tipo_moradia": np.random.choice(["Aluguel", "Propria", "Financiada"], 100)
})

# Inserindo alguns valores ausentes
dados.loc[np.random.choice(dados.index, 5, replace=False), "salario_anual"] = np.nan

# Criando o alvo (0 = baixo risco, 1 = alto risco)
dados["risco_inadimplencia"] = (
    (dados["salario_anual"].fillna(0) < 60000) |
    (dados["valor_emprestimo"] > 80000)
).astype(int)

# 2. Separando os dados
X = dados.drop("risco_inadimplencia", axis=1)
y = dados["risco_inadimplencia"]

X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Pré-processamento
colunas_num = ["idade", "salario_anual", "anos_empregado", "valor_emprestimo"]
colunas_cat = ["tipo_moradia"]

num_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="mean")),
    ("scaler", StandardScaler())
])

cat_pipeline = Pipeline([
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

preprocessador = ColumnTransformer([
    ("num", num_pipeline, colunas_num),
    ("cat", cat_pipeline, colunas_cat)
])

# 4. Modelo
modelo = Pipeline([
    ("preprocessador", preprocessador),
    ("logreg", LogisticRegression(max_iter=1000))
])

# 5. Treinamento
modelo.fit(X_treino, y_treino)

# 6. Avaliação
y_pred = modelo.predict(X_teste)
acuracia = accuracy_score(y_teste, y_pred)

print(f"Acurácia do modelo: {acuracia:.2f}")

# 7. Teste com novo cliente
novo_cliente = pd.DataFrame({
    "idade": [40],
    "salario_anual": [55000],
    "anos_empregado": [5],
    "valor_emprestimo": [90000],
    "tipo_moradia": ["Aluguel"]
})

previsao = modelo.predict(novo_cliente)[0]
print(f"Previsão para o novo cliente: {'Alto Risco' if previsao == 1 else 'Baixo Risco'}")

<img width="561" height="75" alt="Image" src="https://github.com/user-attachments/assets/e2310a3e-5f89-4ae2-bf3a-63ffac9c7032" />


# PROJETO 2 - Previsão de Consumo de Energia

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

print("--- PROJETO 2: Previsão de Consumo de Energia ---")

# 1. Criando o dataset
np.random.seed(42)

datas = pd.date_range(start="2024-01-01", periods=200)
temperaturas = np.random.uniform(10, 35, 200)
dias_uteis = np.random.choice([0, 1], size=200)

df = pd.DataFrame({
    "data": datas,
    "temperatura_media": temperaturas,
    "dia_util": dias_uteis
})

# Alvo: consumo de energia (em kWh)
# Simulando: consumo aumenta com temperatura e dias úteis
df["consumo_energia_kwh"] = (
    2000 + (df["temperatura_media"] * 50) + (df["dia_util"] * 400) + np.random.normal(0, 200, 200)
)

# 2. Engenharia de Atributos
df["mes"] = df["data"].dt.month
df["dia_da_semana"] = df["data"].dt.dayofweek
df["dia_do_ano"] = df["data"].dt.dayofyear
df = df.drop("data", axis=1)

# 3. Separando variáveis
X = df.drop("consumo_energia_kwh", axis=1)
y = df["consumo_energia_kwh"]

X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 4. Pipeline
modelo = Pipeline([
    ("scaler", StandardScaler()),
    ("rf", RandomForestRegressor(random_state=42))
])

# 5. Treinamento
modelo.fit(X_treino, y_treino)

# 6. Avaliação
y_pred = modelo.predict(X_teste)

mae = mean_absolute_error(y_teste, y_pred)
r2 = r2_score(y_teste, y_pred)

print(f"Erro Médio Absoluto (MAE): {mae:.2f} kWh")
print(f"R² Score: {r2:.2f}")

# 7. Teste com um novo dia
novo_dado = pd.DataFrame({
    "temperatura_media": [30],
    "dia_util": [1],
    "mes": [8],
    "dia_da_semana": [2],
    "dia_do_ano": [220]
})

previsao = modelo.predict(novo_dado)[0]
print(f"Previsão de consumo para o novo dia: {previsao:.2f} kWh")

<img width="486" height="87" alt="Image" src="https://github.com/user-attachments/assets/0b903980-c018-4987-9a09-73d42aa20094" />

# PROJETO 3 - Detecção de E-mails Spam

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline

print("--- PROJETO 3: Detecção de E-mails Spam ---")

# 1. Dataset
dados = pd.DataFrame({
    "texto": [
        "oferta imperdível clique aqui agora",
        "ganhe dinheiro fácil e rápido",
        "promoção exclusiva para você",
        "relatório de vendas anexo",
        "oi, tudo bem? reunião amanhã",
        "seguem os documentos solicitados",
        "compre agora e leve grátis",
        "sorteio de prêmios incríveis",
        "favor revisar o contrato enviado",
        "bom dia, reunião confirmada"
    ],
    "categoria": [
        "spam", "spam", "spam",
        "ham", "ham", "ham",
        "spam", "spam", "ham", "ham"
    ]
})

# 2. Divisão dos dados
X = dados["texto"]
y = dados["categoria"]
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Pipeline com TF-IDF e Naive Bayes
modelo = Pipeline([
    ("vetor", TfidfVectorizer()),
    ("nb", MultinomialNB())
])

# 4. Treino e avaliação
modelo.fit(X_treino, y_treino)
acuracia = modelo.score(X_teste, y_teste)

# 5. Teste com novo e-mail
novo_email = ["parabéns! você ganhou um prêmio exclusivo"]
previsao = modelo.predict(novo_email)[0]

print(f"Acurácia: {acuracia:.2f}")
print(f"Novo e-mail classificado como: {previsao}")

<img width="587" height="71" alt="Image" src="https://github.com/user-attachments/assets/a1ef3841-cf09-4ce7-b6e9-62770ac0eb29" />


# PROJETO 4 - Classificação de Espécies de Flores Íris

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

print("--- PROJETO 4: Classificação de Espécies de Flores Íris ---")

# 1. Carregando o dataset
iris = load_iris()
X = iris.data
y = iris.target

# 2. Divisão treino/teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Pipeline: StandardScaler + KNN
modelo = Pipeline([
    ("scaler", StandardScaler()),
    ("knn", KNeighborsClassifier(n_neighbors=5))
])

# 4. Treinamento
modelo.fit(X_treino, y_treino)

# 5. Avaliação
y_pred = modelo.predict(X_teste)
acuracia = accuracy_score(y_teste, y_pred)

print(f"Acurácia do modelo: {acuracia:.2f}")

# 6. Teste com uma nova flor
novo_exemplo = [[5.1, 3.5, 1.4, 0.2]]
predicao = modelo.predict(novo_exemplo)[0]
print(f"Nova flor classificada como: {iris.target_names[predicao]}")

<img width="560" height="67" alt="Image" src="https://github.com/user-attachments/assets/d28c028d-9c20-4177-b615-6a3831afaae1" />


# PROJETO 5 - Previsão de Qualidade de Vinho

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVR
from sklearn.metrics import mean_absolute_error

print("--- PROJETO 5: Previsão de Qualidade de Vinho ---")

# 1. Carregando o dataset
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv'
dados = pd.read_csv(url, sep=';')

# 2. Separando features e alvo
X = dados.drop("quality", axis=1)
y = dados["quality"]

# 3. Divisão treino/teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X, y, test_size=0.3, random_state=42)

# 4. Pipeline: StandardScaler + SVR
modelo = Pipeline([
    ("scaler", StandardScaler()),
    ("svr", SVR())
])

# 5. Treinamento
modelo.fit(X_treino, y_treino)

# 6. Avaliação
y_pred = modelo.predict(X_teste)
mae = mean_absolute_error(y_teste, y_pred)

print(f"Erro Médio Absoluto (MAE): {mae:.2f}")

# 7. Teste com um novo vinho
novo_vinho = pd.DataFrame([X_teste.iloc[0]])
previsao = modelo.predict(novo_vinho)[0]

print(f"Previsão de qualidade para o vinho de teste: {previsao:.2f}")


<img width="484" height="72" alt="Image" src="https://github.com/user-attachments/assets/04ca7c08-821e-436d-835f-0434e84d7ec1" />
